### DISCUSSION TITLE
Providers that must be overridden

### DISCUSSION THREAD

**Discussion by scopendo** [rrousselGit/riverpod#1147] [Ideas]
Hi – apps often need some providers that don't require a create function as they are expected to always have a value supplied via a `ProviderScope` override.  To reduce boilerplate, for basic state types I use a `MustOverrideProvider`:

```dart
/// A [Provider] whose value must be set at startup by overriding the
/// value on [ProviderScope].
///
/// ignore: non_constant_identifier_names
Provider<T> MustOverrideProvider<T>() {
  return Provider<T>(
    (_) => throw ProviderNotOverriddenException(),
  );
}

class ProviderNotOverriddenException implements Exception {
  @override
  String toString() {
    return 'The value for this provider must be set by an override on ProviderScope';
  }
}
```

I wonder if it would be useful to have this included in Riverpod itself, either by making the create function optional or preferably more explicitly via a modifier as with `family` and `autoDispose`?  For example:

```dart
final currentUserPod = StateNotifierProvider.mustOverride<CurrentUserController, UserData>();

class CurrentUserController extends StateNotifier<UserData> {
  CurrentUserController(UserData user) : super(user);
  ...
}
```

**Comment by rrousselGit** [comment]
I'm not sure the value is worth the added confusion of having a larger API surface
